File: colorcam.cpp
// License: Apache 2.0. See LICENSE file in root directory.
// Copyright(c) 2017 Intel Corporation. All Rights Reserved.

/* Include the librealsense C header files */
#include <librealsense2/rs.hpp>
#include <librealsense2/h/rs_pipeline.h>
#include <librealsense2/h/rs_frame.h>
#include <linux/videodev2.h>

#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>
#include <sys/ioctl.h>
#include <unistd.h>
#include <string.h>
#include <fcntl.h>
#include <assert.h>

#define ROUND_UP_2(num)  (((num)+1)&~1)
#define ROUND_UP_4(num)  (((num)+3)&~3)
#define ROUND_UP_8(num)  (((num)+7)&~7)
#define ROUND_UP_16(num) (((num)+15)&~15)
#define ROUND_UP_32(num) (((num)+31)&~31)
#define ROUND_UP_64(num) (((num)+63)&~63)

#define VIDEO_DEVICE "/dev/video8"

# define FRAME_WIDTH 640
# define FRAME_HEIGHT 480
# define STREAM_INDEX 0
# define FPS 30

# define FRAME_FORMAT V4L2_PIX_FMT_YUYV

static int debug=0;

int format_properties(const unsigned int format,
        const unsigned int width,
        const unsigned int height,
        size_t*linewidth, size_t*framewidth)
        {
    size_t lw, fw;
        lw = (ROUND_UP_2 (width) * 2);
        fw = lw * height;

      if(linewidth)*linewidth=lw;
      if(framewidth)*framewidth=fw;

      return 1;
}

void print_format(struct v4l2_format*vid_format) {
  printf("  vid_format->type                =%d\n", vid_format->type );
  printf("  vid_format->fmt.pix.width       =%d\n", vid_format->fmt.pix.width );
  printf("  vid_format->fmt.pix.height      =%d\n", vid_format->fmt.pix.height );
  printf("  vid_format->fmt.pix.pixelformat =%d\n", vid_format->fmt.pix.pixelformat);
  printf("  vid_format->fmt.pix.sizeimage   =%d\n", vid_format->fmt.pix.sizeimage );
  printf("  vid_format->fmt.pix.field       =%d\n", vid_format->fmt.pix.field );
  printf("  vid_format->fmt.pix.bytesperline=%d\n", vid_format->fmt.pix.bytesperline );
  printf("  vid_format->fmt.pix.colorspace  =%d\n", vid_format->fmt.pix.colorspace );
}

/* Function calls to librealsense may raise errors of type rs2_error */
rs2_error * e = 0;
void check_error()
{
    if(e)
    {
        printf("rs2_error was raised when calling %s(%s):\n", rs2_get_failed_function(e), rs2_get_failed_args(e));
        printf("    %s\n", rs2_get_error_message(e));
        exit(EXIT_FAILURE);
    }
}

int main(int argc, char**argv)
{
    // This first section relates only to v4l2loopback, no RealSense code here.

    struct v4l2_capability vid_caps;
    struct v4l2_format vid_format;

    size_t framesize;
    size_t linewidth;

    __u8*buffer;

    const char*video_device=VIDEO_DEVICE;
    int fdwr = 0;
    int ret_code = 0;

    int i;

    if(argc>1) {
        video_device=argv[1];
        printf("using output device: %s\n", video_device);
    }

    fdwr = open(video_device, O_RDWR);
    assert(fdwr >= 0);

    ret_code = ioctl(fdwr, VIDIOC_QUERYCAP, &vid_caps);
    assert(ret_code != -1);

    memset(&vid_format, 0, sizeof(vid_format));

    ret_code = ioctl(fdwr, VIDIOC_G_FMT, &vid_format);
    if(debug)print_format(&vid_format);

    vid_format.type = V4L2_BUF_TYPE_VIDEO_OUTPUT;
    vid_format.fmt.pix.width = FRAME_WIDTH;
    vid_format.fmt.pix.height = FRAME_HEIGHT;
    vid_format.fmt.pix.pixelformat = FRAME_FORMAT;
    vid_format.fmt.pix.sizeimage = framesize;
    vid_format.fmt.pix.field = V4L2_FIELD_NONE;
    vid_format.fmt.pix.bytesperline = linewidth;
    vid_format.fmt.pix.colorspace = V4L2_COLORSPACE_SRGB;

    if(debug)print_format(&vid_format);
    ret_code = ioctl(fdwr, VIDIOC_S_FMT, &vid_format);

    assert(ret_code != -1);

    if(debug)printf("frame: format=%d\tsize=%lu\n", FRAME_FORMAT, framesize);
    print_format(&vid_format);

    if(!format_properties(vid_format.fmt.pix.pixelformat,
                                                vid_format.fmt.pix.width, vid_format.fmt.pix.height,
                                                &linewidth,
                                                &framesize)) {
        printf("unable to guess correct settings for format '%d'\n", FRAME_FORMAT);
    }
    buffer=(__u8*)malloc(sizeof(__u8)*framesize);

    memset(buffer, 0, framesize);

    write(fdwr, buffer, framesize);

    // Now the RealSense code.

    rs2::config cfg;
    cfg.enable_stream(RS2_STREAM_COLOR, FRAME_WIDTH, FRAME_HEIGHT, RS2_FORMAT_YUYV, FPS);

    rs2::pipeline pipe;
    pipe.start(cfg);

    printf("Copying frame data to loopback device.\nCTRL-C to quit.\n");
    for(;;)
    {
        rs2::frameset frames = pipe.wait_for_frames();
        rs2::frame frame = frames.first(RS2_STREAM_COLOR);
        if (frame)
            buffer = (__u8*)frame.get_data();
            write(fdwr, buffer, framesize);
    }

    return EXIT_SUCCESS;
}

File: colorcam.py

#coding:utf-8

import pyrealsense2 as rs
import fcntl, sys, os
from v4l2 import *
import time
import numpy as np
import cv2
import sys

def start_camera():
	devName = '/dev/video80'
	if len(sys.argv) >= 2:
		devName = sys.argv[1]
	width = 640
	height = 480
	if not os.path.exists(devName):
		print("Warning: device does not exist",devName)
	device = open(devName, 'wb+', buffering=0)

	print(device)
	capability = v4l2_capability()
	print("get capabilities result", (fcntl.ioctl(device, VIDIOC_QUERYCAP, capability)))
	print("capabilities", hex(capability.capabilities))

	fmt = V4L2_PIX_FMT_YUYV

	print("v4l2 driver: ", capability.driver)
	format = v4l2_format()
	format.type = V4L2_BUF_TYPE_VIDEO_OUTPUT
	format.fmt.pix.pixelformat = fmt
	format.fmt.pix.width = width
	format.fmt.pix.height = height
	format.fmt.pix.field = V4L2_FIELD_NONE
	format.fmt.pix.bytesperline = width * 2
	format.fmt.pix.sizeimage = width * height * 2
	format.fmt.pix.colorspace = V4L2_COLORSPACE_JPEG

	print("set format result", (fcntl.ioctl(device, VIDIOC_S_FMT, format)))
	#Note that format.fmt.pix.sizeimage and format.fmt.pix.bytesperline 
	#may have changed at this point

	pipeline = rs.pipeline()
	config = rs.config()
	config.enable_stream(rs.stream.depth, 640, 480, rs.format.z16, 30)
	# config.enable_stream(rs.stream.color, 640, 480, rs.format.yuyv, 30)
	config.enable_stream(rs.stream.color, 640, 480, rs.format.rgb8, 30)
	
	pipeline.start(config)

	while True:
		frames = pipeline.wait_for_frames()
		depth_frame = frames.get_depth_frame()
		color_frame = frames.get_color_frame()
		color_data = color_frame.get_data()

		color_nparray = np.asanyarray(color_data)
		depth_nparray= np.asanyarray(depth_frame.get_data())

        # convert rgb to yuyv
		yuv = cv2.cvtColor(color_nparray, cv2.COLOR_RGB2YUV)
		buf = np.zeros((format.fmt.pix.height, 2*format.fmt.pix.width), dtype=np.uint8)
		for i in range(format.fmt.pix.height):
			buf[i,::2] = yuv[i,:,0]
			buf[i,1::4] = yuv[i,::2,1]
			buf[i,3::4] = yuv[i,::2,2] 

		device.write(buf.tostring())
        

if __name__=="__main__":
	start_camera()

